File: crane/environments.py
from crane.webserver import app
from flask import jsonify, request
from Backend.Environments import EnvironmentProvider
from Backend.Deployer import Deployer
from Backend.HostProvider import HostProvider
from crane.Backend.Utils.SSHConnection import SSHConnection

host_provider = HostProvider(SSHConnection())
environment_provider = EnvironmentProvider(host_provider)

@app.route("/environment", methods=['GET'])
def get_environments():
    return jsonify(result=environment_provider.get_environments())

@app.route("/environment", methods=['POST'])
def add_environment():
    json = request.get_json()
    env_id = environment_provider.add_environment(json['name'])
    return jsonify(id=env_id)

@app.route("/environment/<id>/host", methods=['POST'])
def add_host_to_environment(id):
    json = request.get_json()
    environment_provider.add_host_to_environment(id, json['id'])
    return ""

@app.route("/environment/<id>/host/<hostid>", methods=['DELETE'])
def delete_host_from_environment(id, hostid):
    environment_provider.delete_host_from_environment(id, hostid)
    return ""

@app.route("/environment/<id>", methods=['DELETE'])
def delete_environment(id):
    environment_provider.delete_environment(id)
    return ""

@app.route("/environment/<id>/deploy", methods=['POST'])
def deploy_in_environment(id):
    data = request.get_json()
    count = data['count']
    hosts = environment_provider.get_hosts_for_deployment(id, count)
    deployer = Deployer(host_provider)
    name = data['container']['name']
    count = 1
    for host in hosts:
        data['container']['name'] = "{0}-{1}".format(name, count)
        result = deployer.deploy(host, data)
        count = count + 1
    return ""

File: crane/js/environments.js
(function(){

angular.module('crane')

.controller('EnvironmentsController', EnvironmentsController);

EnvironmentsController.$inject = ['$scope', '$http', '$modal'];

function EnvironmentsController($scope, $http, $modal) {
  $scope.add_environment = { 'active':false, 'environment' : {} }

  $scope.load_environments = load_environments;
  $scope.start_add_environment = start_add_environment;
  $scope.cancel_add_environment = cancel_add_environment;
  $scope.do_add_environment = do_add_environment;
  $scope.start_add_host = start_add_host;
  $scope.cancel_add_host = cancel_add_host;
  $scope.do_add_host = do_add_host;
  $scope.load_hosts = load_hosts;
  $scope.remove_host = remove_host;
  $scope.remove_environment = remove_environment;

  function load_hosts() {
    $http.get("/host").success(function(data, status) {
      $scope.hosts = data.result;
    });
  }

  function load_environments() {
    $http.get("/environment").success(function(data, status) {
      $scope.environments = data.result;
    });
  }

  function do_add_environment() {
     $http.post("/environment", $scope.add_environment.environment).success(function(data) {
        $scope.load_environments();
     });
     $scope.add_environment.active = false;
  };

  function do_add_host(environment) {
     $http.post("/environment/" + String(environment.id) + "/host", {'id': environment.add_host.id} ).success(function(data) {
        $scope.load_environments();
     });
     environment.add_active = false;
  };

  function start_add_environment() { $scope.add_environment.active = true; };

  function start_add_host(environment) { environment.add_active = true; };

  function cancel_add_environment() { $scope.add_environment.active = false; };

  function cancel_add_host(environment) { environment.add_active = false; };

  function remove_host(environment, host) {
     $http.delete("/environment/" + String(environment.id) + "/host/" + String(host.id)).success(function(data) {
       $scope.load_environments();
     });
  }

  function remove_environment(environment) {
     $http.delete("/environment/" + String(environment.id)).success(function(data) {
       $scope.load_environments();
     });
  }

  $scope.load_environments();
  $scope.load_hosts();
}

})();
